nixos/modules/tasks/auto-upgrade.nix

```
@@ -174,6 +253,251 @@ in
+              };
+
+              options.pkgs = lib.mkOption {
+                readOnly = true;
+                description = ''
+                  The last known-working version of nixpkgs that passes all `requireBuilds`.
+
+                  This option is read-only, and is the result of this package.
+                '';
+                type = lib.types.pkgs;
+              };
+
+              # Internal options and configuration, to make this work
+              options.desyncUpgradeScript = lib.mkOption {
+                # Derivation called to upgrade the desync files.
+                # `$out` will be symlinked from `/var/lib/nixos/desyncs/${name}.nix by the `upgradeScript`
+                readOnly = true;
+                internal = true;
+                visible = false;
+                description = ''
+                  Derivation called to upgrade the desync files.
+                  `$out` will be symlinked from `/var/lib/nixos/desyncs/${name}.nix by the `upgradeScript`
+                '';
+                type = lib.types.package;
+              };
+
+              options.upgradeFailedScript = lib.mkOption {
+                # Derivation called when the latest upgrade failed.
+                # `$out` will be symlinked from `/var/lib/nixos/desyncs/${name}.nix by the `upgradeScript`
+                readOnly = true;
+                internal = true;
+                visible = false;
+                description = ''
+                  Derivation called when the latest upgrade failed.
+                  `$out` will be symlinked from `/var/lib/nixos/desyncs/${name}.nix by the `upgradeScript`
+                '';
+                type = lib.types.package;
+              };
+
+              config =
+                let
+                  desync =
+                    if builtins.pathExists /var/lib/nixos/desyncs/${name}.nix then
+                      import /var/lib/nixos/desyncs/${name}.nix
+                    else
+                      {
+                        # If we have no desync saved, it means that any configuration that builds
+                        # would necessarily be with the default path.
+                        # This can happen eg. on the `nixos-rebuild` that adds the desync configuration
+                        # to a nixos config, that would most likely not happen during an auto-upgrade.
+                        desyncAge = 0;
+                        path = config.defaultPath;
+                      };
+                in
+                {
+                  inherit (desync) desyncAge path;
+                  pkgs = config.pkgsFor desync.path;
+
+                  desyncUpgradeScript = pkgs.runCommand "nixos-desync-${name}" { } ''
+                    #!/bin/sh
+                    set -euo pipefail
+
+                    # Force dependencies on `requireBuilds`
+                    # ${lib.concatStringsSep " " (config.requireBuilds (config.pkgsFor (config.defaultPath)))}
```

Up to you, but sometimes these deeply nested function applications can be easier to read as a pipe:

```suggestion
                    # ${lib.pipe config.defaultPath [
                      config.pkgsFor
                      config.requireBuilds
                      (lib.concatStringsSep " ")
                    ]}
```

Reading it again makes me question whether `pkgsFor` and `requireBuilds` are actually functions, though? Just based on their naming convention, I would assume `requireBuilds` is a boolean, and `pkgsFor` is either a map/dict attrset or a function of type `desync -> nixpkgs-instance`.

---

One day... one day we'll have [a $ operator](https://github.com/NixOS/nix/issues/1845). But I don't have enough energy to start a new RFC now so 🤷 

As for the question about pkgsFor and requireBuilds, they must take the package set as a parameter. The rationale here is, that the desync system will try bumping the channel, run these two functions, if everything builds fine then accept the new version, and if not actually desync from the channel.

Technically speaking they're always used with `requireBuilds (pkgsFor path)`, but I have `pkgsFor` into a separate option so that it can have the nixpkgs default, which should be more convenient in most cases.

Does that make sense to you?

---

> One day... one day we'll have [a $ operator](https://github.com/NixOS/nix/issues/1845).

We are at least [getting](https://github.com/NixOS/rfcs/pull/148) a `|>` pipe operator :tada: